Add support to collect PGO profile on iOS.

dewei-zhu opened the pull request:
#### 20e6698edf478e045292cb46fcc9ac102dd7003a
<pre>
Add support to collect PGO profile on iOS.
<a href="https://rdar.apple.com/138666783">rdar://138666783</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=288438">https://bugs.webkit.org/show_bug.cgi?id=288438</a>

Reviewed by Per Arne Vollan, Alexey Proskuryakov, and Wenson Hsieh.

Implement a way that defers the start of PGO collection until app container tmp directory access is granted.
Allow &apos;F_GETFL&apos; for GPU process which is required by llvm profile.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/runtime/InitializeThreading.cpp:
* Source/JavaScriptCore/tools/LLVMProfiling.cpp: Added.
(JSC::initializeLLVMProfiling):
* Source/JavaScriptCore/tools/LLVMProfiling.h: Added.
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/LLVMProfilingUtils.h: Added.
(WTF::initializeLLVMProfiling):
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/ScriptController.cpp:
* Source/WebCore/platform/LLVMProfiling.cpp: Added.
(WebCore::initializeLLVMProfiling):
* Source/WebCore/platform/LLVMProfiling.h: Added.
* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::initializeGPUProcess):
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::addWebsiteDataStore):
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.GPU.sb.in:
* Source/WebKit/Shared/AuxiliaryProcess.cpp:
(WebKit::AuxiliaryProcess::grantAccessToContainerTempDirectory):
* Source/WebKit/Shared/AuxiliaryProcess.h:
* Source/WebKit/Shared/Cocoa/WebKit2InitializeCocoa.mm:
* Source/WebKit/Shared/LLVMProfiling.h: Added.
(WebKit::initializeLLVMProfiling):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::platformSetWebsiteDataStoreParameters):

Canonical link: <a href="https://commits.webkit.org/293649@main">https://commits.webkit.org/293649@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e47f5c149f48327801d978e255eeb1a51299fbf3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99507 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19157 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9411 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104638 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50109 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101548 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19445 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27590 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75740 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/32840 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102514 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14808 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89851 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56099 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/14605 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/7836 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49468 "Built successfully") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/92190 "Build is in progress. Recent messages:") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84537 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/7922 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106996 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/98126 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26621 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/19431 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/84700 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26983 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86055 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/84217 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21368 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28895 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6591 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/20408 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26561 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31762 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/121742 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26381 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34001 "Found 1 new JSC stress test failure: ChakraCore.yaml/ChakraCore/test/es6/ES6MathAPIs.js.default (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29694 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27948 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Source/JavaScriptCore/runtime/InitializeThreading.cpp:
```diff
@@ -61,8 +61,13 @@ WTF_IGNORE_WARNINGS_IN_THIRD_PARTY_CODE_BEGIN
 #if ENABLE(LLVM_PROFILE_GENERATION)
+#if PLATFORM(IOS_FAMILY)
+#include <wtf/LLVMProfileUtils.h>
+extern "C" char __llvm_profile_filename[] = "%t/WebKitPGO/JavaScriptCore_%m_pid%p%c.profraw";
```
Would it be feasible to use `%t` on macOS, too?

dewei-zhu replied:
It should, but that will require collection script be able to resolve `%t` as well and I'm not certain if all `$TMPDIR` variables for harness script and Safari are always the same. Also, using `%t` would also make `Safari.app` macOS to start dumping PGO profiles which current implementation blocked it by limiting Safari.app access to /tmp in sandbox rule.

aproskuryakov commented on Source/WTF/wtf/LLVMProfileUtils.h:
```diff
@@ -0,0 +1,48 @@
+}
+
+template<typename OnceFlagDiversifier>
+ALWAYS_INLINE void setupProfileCollection(const char *name)
```
"Set up" is two words, so this should be `setUpProfileCollection`.

aproskuryakov commented on Source/JavaScriptCore/runtime/InitializeThreading.cpp:
```diff
@@ -144,6 +149,10 @@ void initialize()
         g_jscConfig.initializeHasBeenCalled = true;
+
+#if ENABLE(LLVM_PROFILE_GENERATION) && PLATFORM(IOS_FAMILY)
+        WTF::setupProfileCollection<JSCProfileTag>("JavaScriptCore");
```
Why is this done last, not first? Even though profiling initialization code is likely useless in practice, it seems more consistent with the goal than artificially skipping it.

emw-apple commented on Source/WebCore/platform/CommonAtomStrings.cpp:
```diff
@@ -44,6 +55,10 @@ void initializeCommonAtomStrings()
+
+#if ENABLE(LLVM_PROFILE_GENERATION) && PLATFORM(IOS_FAMILY)
+    WTF::setupProfileCollection<WebCoreProfileTag>("WebCore");
+#endif
```
This seems like a weird place to do initialization, since it's not a general-purpose initialization function, it's a specific part of the atom strings implementation. I can see why it's useful but it would be nice to have put this somewhere more hygenic.

whsieh replied:
+1. Maybe you could check where it (previously) was added, when we used to use dispatch notifications for PGO collection?

dewei-zhu replied:
No longer doing this in new implementation.

emw-apple commented on Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
```diff
@@ -606,10 +611,13 @@ static void setVideoDecoderBehaviors(OptionSet<VideoDecoderBehavior> videoDecode
         SandboxExtension::consumePermanently(*handle);
+#if ENABLE(LLVM_PROFILE_GENERATION)
+        WTF::sendInitializeProfileCollectionNotification();
+#endif
```
Ditto. It's not clear to me that this function is even called during initialization. Why did you choose to make the call from here, and why only for iOS?

dewei-zhu replied:
> Ditto. It's not clear to me that this function is even called during initialization. Why did you choose to make the call from here, and why only for iOS?

This is the function that gets called when sandbox permission is granted to tell the process to start PGO profile collection.
Note, this is not `WTF::setupProfileCollection`,

emw-apple replied:
It looks like it has lots of other uses. I see this function is called as part of the `SetWebsiteDataStoreParameters` xpc endpoint, and call be invoked arbitrarily from the UI process via `WebProcessProxy::setWebsiteDataStore()`. So, neither the name nor the usage pattern in the codebase imply "start PGO profile collection" to me.

dewei-zhu replied:
I can change the name to `startProfileCollection` to be more descriptive.
This will only be invoked when we grant sandbox tmp dir access (note it's within `if (auto& handle = parameters.containerTemporaryDirectoryExtensionHandle)`).
Furthermore, it will be a no-op once the PGO profile is created, as we do `notify_cancel(token)` on a successful profile creation.

dewei-zhu replied:
No longer doing this in new implementation.
However, on macOS, there is no container access grant and we don't need to defer PGO profile collection.

emw-apple commented on Source/WebKit/NetworkProcess/NetworkProcess.cpp:
```diff
@@ -536,10 +540,13 @@ void NetworkProcess::addWebsiteDataStore(WebsiteDataStoreParameters&& parameters
         SandboxExtension::consumePermanently(*handle);
+#if ENABLE(LLVM_PROFILE_GENERATION)
+        WTF::sendInitializeProfileCollectionNotification();
+#endif
```
Ditto.

dewei-zhu replied:
No longer doing this in new implementation.

emw-apple reviewed (commented):
Nice!

re: use of `PLATFORM(IOS_FAMILY)` across this patch: I don't see why we should have separate code paths for iOS and macOS PGO collection, and I think it's going to be easier for us all to have a uniform behavior on all platforms. Have you tested the new implementation on macOS and seen a regression?

whsieh commented on Source/WTF/wtf/LLVMProfileUtils.h:
```diff
@@ -0,0 +1,48 @@
+
+#if ENABLE(LLVM_PROFILE_GENERATION) && PLATFORM(IOS_FAMILY)
+#include "text/MakeString.h"
+#include "text/WTFString.h"
```
```suggestion
#include <wtf/text/MakeString.h>
#include <wtf/text/WTFString.h>
```

…for consistency with other WTF code.

whsieh commented on Source/WTF/wtf/LLVMProfileUtils.h:
```diff
@@ -0,0 +1,48 @@
+        int status = notify_register_dispatch(notificationName, &token, dispatch_get_main_queue(), ^(int) {
+            __llvm_profile_initialize_file();
+            const char *filename = __llvm_profile_get_filename();
+            if (!access(filename, F_OK))
```
Nit - not sure the local variable adds much here.

```suggestion
            if (!access(__llvm_profile_get_filename(), F_OK))
```

whsieh commented on Source/WTF/wtf/LLVMProfileUtils.h:
```diff
@@ -0,0 +1,48 @@
+    static std::once_flag registerFlag;
+    std::call_once(registerFlag, [name] {
+        int token;
+        auto notificationName = profileCollectionNotificationNameForCurrentProcess().utf8().data();
```
```suggestion
        auto* notificationName = profileCollectionNotificationNameForCurrentProcess().utf8().data();
```

(You could also just inline this below)

pvollan commented on Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.GPU.sb.in:
```diff
@@ -28,6 +28,7 @@
     (fcntl-command
         F_ADDFILESIGS_RETURN
         F_CHECK_LV
+        F_GETFL
```
The sandbox change looks good.

pvollan commented on Source/WTF/wtf/LLVMProfilingUtils.h:
```diff
@@ -0,0 +1,58 @@
+        const char *profilePath = __llvm_profile_get_filename();
+        int pid = getpid();
+        if (access(profilePath, F_OK))
+            RELEASE_LOG(LLVMProfiling, "Process(%d) failed to create LLVM profile at %s.", pid, profilePath);
```
RELEASE_LOG_ERROR?

pvollan commented on Source/WTF/wtf/LLVMProfilingUtils.h:
```diff
@@ -0,0 +1,58 @@
+        if (access(profilePath, F_OK))
+            RELEASE_LOG(LLVMProfiling, "Process(%d) failed to create LLVM profile at %s.", pid, profilePath);
+        else
+            RELEASE_LOG_ERROR(LLVMProfiling, "Process(%d) created LLVM profile at %s.", pid, profilePath);
```
Should the error log be on the log statement above instead?

pvollan reviewed (approved):
R=me.

aproskuryakov commented on Source/WebKit/GPUProcess/GPUProcess.cpp:
```diff
@@ -249,6 +252,13 @@ void GPUProcess::initializeGPUProcess(GPUProcessCreationParameters&& parameters,
+#if ENABLE(LLVM_PROFILE_GENERATION) && PLATFORM(IOS_FAMILY)
+    WebCore::initializeLLVMProfiling();
+    WebKit::initializeLLVMProfiling();
+    JSC::initializeLLVMProfiling();
```
The order is strange here, as WebCore is between WebKit and JSC.

dewei-zhu replied:
Will fix that.

aproskuryakov commented on Source/JavaScriptCore/tools/LLVMProfiling.cpp:
```diff
@@ -0,0 +1,42 @@
+#include "config.h"
+#include "LLVMProfiling.h"
+
+#if ENABLE(LLVM_PROFILE_GENERATION)
```
Much of this PR looks like it adds code that's shared between iOS and macOS, but it's actually iOS-only. This is going to be confusing. Perhaps there is a way to refactor to share of the lower level more code (I understand that some of the behavior is different intentionally).

dewei-zhu committed:
Add support to collect PGO profile on iOS.
rdar://138666783
https://bugs.webkit.org/show_bug.cgi?id=288438

Reviewed by Per Arne Vollan, Alexey Proskuryakov, and Wenson Hsieh.

Implement a way that defers the start of PGO collection until app container tmp directory access is granted.
Allow 'F_GETFL' for GPU process which is required by llvm profile.

* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/runtime/InitializeThreading.cpp:
* Source/JavaScriptCore/tools/LLVMProfiling.cpp: Added.
(JSC::initializeLLVMProfiling):
* Source/JavaScriptCore/tools/LLVMProfiling.h: Added.
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/LLVMProfilingUtils.h: Added.
(WTF::initializeLLVMProfiling):
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/ScriptController.cpp:
* Source/WebCore/platform/LLVMProfiling.cpp: Added.
(WebCore::initializeLLVMProfiling):
* Source/WebCore/platform/LLVMProfiling.h: Added.
* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::initializeGPUProcess):
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::addWebsiteDataStore):
* Source/WebKit/Resources/SandboxProfiles/ios/com.apple.WebKit.GPU.sb.in:
* Source/WebKit/Shared/AuxiliaryProcess.cpp:
(WebKit::AuxiliaryProcess::grantAccessToContainerTempDirectory):
* Source/WebKit/Shared/AuxiliaryProcess.h:
* Source/WebKit/Shared/Cocoa/WebKit2InitializeCocoa.mm:
* Source/WebKit/Shared/LLVMProfiling.h: Added.
(WebKit::initializeLLVMProfiling):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::WebProcess::platformSetWebsiteDataStoreParameters):

Canonical link: https://commits.webkit.org/293649@main
